ts/nix.ts

```diff
@@ -132,37 +151,148 @@ export function nixStrLit(
   s: TemplateStringsArray | string,
   ...interpolations: Array<NixStrLitInterpolatable>
 ): NixExpression {
-  if (typeof s === "string") return nixStrLit`${s}`;
-  const escape = (str: string) =>
-    ["\\", '"', "$"].reduce(
-      (str, char) => str.replaceAll(char, `\\${char}`),
-      str,
-    );
-  const escapedTemplate = s.map(escape);
-  const rawNixExpressionString =
-    '"' +
-    escapedTemplate.reduce((acc, part, i) => {
-      const interpolation = interpolations[i];
-      switch (typeof interpolation) {
-        case "string":
-          return acc + part + escape(interpolation);
-        case "object":
-          if ("rawNixExpressionString" in interpolation) {
-            return (
-              acc + part + "${" + interpolation.rawNixExpressionString + "}"
-            );
+  if (typeof s === "string") {
+    return { __ast: { type: "strLit", str: interpolatedStringFromString(s) } };
+  }
+  return {
+    __ast: {
+      type: "strLit",
+      str: interpolatedStringFromTemplate(
+        s,
+        interpolations.map((interpolation): NixAst => {
+          if (typeof interpolation === "string") {
+            return {
+              type: "strLit",
+              str: interpolatedStringFromString(interpolation),
+            };
           }
-          return (
-            acc +
-            part +
-            "${" +
-            interpolation.nixExpression.rawNixExpressionString +
-            "}"
+          if ("__ast" in interpolation) {
+            return interpolation.__ast;
+          }
+          return interpolation.nixExpression.__ast;
+        }),
+      ),
+    },
+  };
+}
+
+/**
+ * Converts a `NixExpression` to a (hopefully concise) human readable string
+ * with incidental dependencies snipped out with "[...]"
+ */
+export function toHumanReadable(nixExpr: NixExpression): string {
+  const astToHumanReadable = (node: NixAst): string => {
+    if (node.type !== "strLit") {
+      return "[...]";
+    }
+    return node.str.rest.reduce(
+      (acc, [node, str]) => acc + astToHumanReadable(node) + str,
+      node.str.initial,
+    );
+  };
+  return astToHumanReadable(nixExpr.__ast);
+}
+
+/**
+ * Converts a `NixExpression` to a string representing that Nix expression.
+ *
+ * It is advised to defer calling this unless actually serializing to disk.
+ * This way Nix expressions remain as the type `NixExpression`.
+ */
+export function toNixString(nixExpr: NixExpression): string {
```

Comment:
```suggestion
export function renderNixExpression(nixExpr: NixExpression): string {
```

I think `toNixString` is confusing, since it could talk about nix string literals.